lib/ht_members.rb

```diff
@@ -1,17 +1,151 @@
+The data may be mocked if you pass it in to new(),
+otherwise it will try to read it from db, and if that fails
+it will use canned data (CANNED_DATA).
+
+@mocked and @canned are set accordingly and readable, so you can
+tell where the data is coming from
+
+Usage:
+
+  htm = HTMembers.new()
+  puts "using canned data? #{htm.canned}"
+  cc = htm.get("yale")["country_code"]
+  wt = htm.get("harvard")["weight"]
+
+Or if you want to provide your own mock data:
+
+  htm = HTMembers.new({
+    "haverford" => {
+      "country_code" => "us",
+      "weight"       => 0.67,
+    }
+  })
+  puts "using canned data? #{htm.canned}"
+  puts "using mocked data? #{htm.mocked}"
+  htm.get("haverford")["country_code"]
+  htm.get("haverford")["weight"]
+
+=end
+
 class HTMembers
+  def self.template(cc, wt, sym)
+    {
+      "country_code" => cc,
+      "weight"       => wt,
+      "oclc_sym"     => sym,
+    }
+  end
 
-  def initialize(*members)
-    @members = members.empty? ? load_from_db : members
+  # If we get more cases like universityofcalifornia then this might
+  # look better as a series of SELECT ... UNION SELECT ....
+  SQL = %w<
+    SELECT DISTINCT 
+    IF(
+      mapto_inst_id='universityofcalifornia',
+      inst_id, 
+      mapto_inst_id
+    ) AS member_id,
+    country_code,
+    weight,
+    oclc_sym
+    FROM ht_institutions
+    WHERE enabled = '1'
+  >.join(' ')
+
+  CANNED_DATA = {
+    "brocku"  => HTMembers.template("ca", 0.67, "BRX"),
+    "harvard" => HTMembers.template("za", 1.33, "HUL"),
+    "umich"   => HTMembers.template("us", 1.33, "EYM"),
+    "yale"    => HTMembers.template("us", 1.33, "YUS"),
+  }
+
+  attr_reader :mocked, :canned, :members
+  def initialize(members = {})
+    @mocked  = !members.empty?
+    @canned  = false
+    @members = @mocked ? members : load_from_db
+  end
+
+  # Are all the ENV vars required for DB connection set?
+  def db_env_set?
+    problem = false
+    %w[HOST USERNAME PASSWORD PORT DATABASE].each do |name_part|
+      name = "MYSQL_#{name_part}";
+      # ENV[name] must be set, non-nil and non-empty.
+      if !ENV.key?(name) || ENV[name].nil? || ENV[name].empty? then
+        warn "ENV[#{name}] not set correctly (#{ENV[name]})"
+        problem = true
+      end
+    end
+
+    return !problem
   end
 
   def load_from_db
-    # do that
+    data = {}
+    # Attempt loading from db, use canned data if it fails.
+    begin
+      mysql_client = Mysql2::Client.new(
+        :host     => ENV["MYSQL_HOST"],
+        :username => ENV["MYSQL_USERNAME"],
+        :password => ENV["MYSQL_PASSWORD"],
+        :port     => ENV["MYSQL_PORT"],
+        :database => ENV["MYSQL_DATABASE"],
+        :connect_timeout => 5,
+      )
+
+      res = mysql_client.query(SQL)
+      res.each do |row|
+        data[row[:member_id]] = HTMembers.template(
+          row["country_code"], row["weight"], row["oclc_sym"]
+        )
+      end
+    rescue Mysql2::Error::ConnectionError => e
```

Comment:
This fall-through to canned data makes me uncomfortable. Am I right that if the DB is down or times out or something, we could end up using canned data when we don't want it and not notice until all sorts of damage had been done to the mongo datastore.